packages/theme-addresses/addressForm.ts

```diff
@@ -0,0 +1,274 @@
+  country: '[name="address[country]"]',
+  zone: '[name="address[province]"]',
+  postalCode: '[name="address[zip]"]',
+  city: '[name="address[city]"]',
+  phone: '[name="address[phone]"]',
+};
+
+interface FormElement {
+  wrapper?: HTMLElement;
+  input?: HTMLInputElement | HTMLSelectElement;
+  labels?: HTMLElement;
+}
+
+interface FormElements {
+  lastName?: FormElement;
+  firstName?: FormElement;
+  company?: FormElement;
+  address1?: FormElement;
+  address2?: FormElement;
+  country?: FormElement;
+  zone?: FormElement;
+  postalCode?: FormElement;
+  city?: FormElement;
+  phone?: FormElement;
+}
+
+interface AddressFormOptions {
+  loaded?(): void;
+}
+
+export default function AddressForm(
+  rootEl: HTMLElement,
+  locale: string = 'en',
+  options: AddressFormOptions = {},
+) {
+  var formElements = loadFormElements(rootEl);
+  validateElements(formElements);
+
+  if (COUNTRIES) {
+    init(rootEl, formElements, options);
+  } else {
+    loadCountries(locale).then(function(countries) {
+      COUNTRIES = countries;
+      init(rootEl, formElements, options);
+    });
+  }
+}
+
+/**
+ * Runs when countries have been loaded
+ */
+function init(rootEl: HTMLElement, formElements: FormElements, options: AddressFormOptions) {
+  populateCountries(formElements);
+  var selectedCountry = formElements.country.input
+    ? formElements.country.input.value
+    : null;
+  handleCountryChange(rootEl, formElements, selectedCountry);
+  if (options.loaded) {
+    options.loaded();
+  }
+}
+
+/**
+ * Handles when a country change: set labels, reorder fields, populate zones
+ */
+function handleCountryChange(rootEl: HTMLElement, formElements: FormElements, countryCode: string) {
+  var country = getCountry(countryCode);
+
+  setEventListeners(rootEl, formElements, country);
+  setLabels(formElements, country);
+  reorderFields(rootEl, formElements, country);
+  populateZones(formElements, country);
+}
+
+/**
+ * Sets up event listener for country change
+ */
+function setEventListeners(rootEl: HTMLElement, formElements: FormElements, country: Country) {
+  formElements.country.input.addEventListener('change', function(event) {
+    handleCountryChange(rootEl, formElements, (event.target as HTMLSelectElement).value);
+  });
+}
+
+/**
+ * Reorder fields in the DOM and add data-attribute to fields given a country
+ */
+function reorderFields(rootEl: HTMLElement, formElements: FormElements, country: Country) {
+  var formFormat = country.formatting.edit;
+  var rowIndex = 0;
+
+  var countryWrapper = formElements.country.wrapper;
+  var afterCountry = false;
+  getOrderedField(formFormat).forEach(function(row) {
+    row.forEach(function(line) {
+      formElements[line].wrapper.dataset.lineCount = row.length;
+      if (!formElements[line].wrapper) {
+        return;
+      }
+      if (line === 'country') {
+        afterCountry = true;
+        return;
+      }
+
+      if (afterCountry) {
+        rootEl.append(formElements[line].wrapper);
```

Comment:
You should create a duplicate of rootEl in memory, append the newly ordered fields to it, and then replace the version of rootEl in the DOM with your updated one. This will significantly reduce the number of DOM operations you are performing.

Reply:
👍

Reply:
Actually in this case it fails on the `insertBefore` bellow because I'm dealing with in memory and in the DOM element I think.
But I only have ~10 operations here so we should be good in terms of performance. I'll change it for the country `select` though.